### PR TITLE
change kpartx separator to '-part'

### LINKED ISSUE #13: create fails mysteriously after mkfs
I have this rather exotic environment here, I admit. I am building a VM with `-t blockdev` and adopting three iSCSI volumes exported from a Dell PowerVault 3200i, using basically this procedure:

```
gnt-instance add \
              -n chi-node-04.torproject.org \
              -o debootstrap+buster \
              -t blockdev --no-wait-for-sync \
              --net 0:ip=pool,network=gnt-chi-01 \
              --no-ip-check \
              --no-name-check \
              --disk 0:adopt=/dev/disk/by-id/dm-name-tb-builder-03-root \
              --disk 1:adopt=/dev/disk/by-id/dm-name-tb-builder-03-swap,name=swap \
              --disk 2:adopt=/dev/disk/by-id/dm-name-tb-builder-03-srv \
              --backend-parameters memory=16g,vcpus=8 \
              tb-builder-03.torproject.org
```

(I spare you the gory details of how the disks are actually created, exported, and shared from the SAN to the ganeti nodes, fun times there.)

For some reason, that doesn't actually work: the `create` hook never runs on adoption. *But* we can coerce it into running by doing a `gnt-instance reinstall`. And *then* the real fun begins. I added tracing in `create` because I couldn't figure out wtf was going on (`set -x` on top). After many attempts (including wiping the `-root` device above), I ended up with this trace:

```
+ '[' msdos = msdos ']'
+ format_disk0 /dev/disk/by-id/dm-name-tb-builder-03-root
+ local ARGS
+ sfdisk --help
+ grep -q -e --cylinders
+ sfdisk --help
+ fgrep -q -- --no-reread
+ ARGS='--no-reread '
+ sfdisk --no-reread --quiet /dev/disk/by-id/dm-name-tb-builder-03-root
Partition #1 contains a ext4 signature.
Re-reading the partition table failed.: Invalid argument
++ map_disk0 /dev/disk/by-id/dm-name-tb-builder-03-root
++ blockdev=/dev/disk/by-id/dm-name-tb-builder-03-root
+++ kpartx -l -p- /dev/disk/by-id/dm-name-tb-builder-03-root
+++ grep -m 1 -- '-1 : .*/dev/disk/by-id/dm-name-tb-builder-03-root'
+++ awk '{print $1}'
++ filesystem_dev_base=tb-builder-03-root-1
++ '[' -z tb-builder-03-root-1 ']'
++ kpartx -a -s -p- /dev/disk/by-id/dm-name-tb-builder-03-root
++ filesystem_dev=/dev/mapper/tb-builder-03-root-1
++ '[' '!' -b /dev/mapper/tb-builder-03-root-1 ']'
++ echo /dev/mapper/tb-builder-03-root-1
+ filesystem_dev=/dev/mapper/tb-builder-03-root-1
+ CLEANUP+=("unmap_disk0 $blockdev")
+ mke2fs -Fjt ext4 /dev/mapper/tb-builder-03-root-1
mke2fs 1.44.5 (15-Dec-2018)
Creating filesystem with 2621184 4k blocks and 655360 inodes
Filesystem UUID: 5ef909ab-bcfe-407d-b2d2-cd4e67fc406a
Superblock backups stored on blocks: 
        32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (16384 blocks): done
Writing superblocks and filesystem accounting information: done

++ /sbin/blkid -o value -s UUID /dev/mapper/tb-builder-03-root-1
+ root_uuid=
+ cleanup
```

for those less familiar with the code, what's happening is that `map_disk0` is correctly finding the new partition and mapping it to `tb-builder-03-root-1`. then `mkfs` comes in and creates the filesystem. but then the `blkid` command comes in and can't find the partition anymore!

indeed, more debugging shows that the device is just gone:

```
Writing superblocks and filesystem accounting information: done

+ sync
+ sleep 3
+ ls -al /dev/mapper/tb-builder-03-root-1
ls: cannot access '/dev/mapper/tb-builder-03-root-1': No such file or directory
+ cleanup
```

after more debugging, what I discovered is that, for a reason I cannot explain, `mkfs` actually changes the device maps, and the partition is *renamed* from `tb-builder-03-root-1` to `tb-builder-03-root-part1` somewhat magically.

i can't figure out what happens, or what the right fix is for this, but this feels like a bug in this bootstrapping mechanism.

### PR DESCRIPTION
**Pull request by jcharaoui**
This is a fix for #13 and supersedes the #14 PR.

### CODE REVIEW

**Review by anarcat — APPROVED** [review]
LGTM.

File: create [hunk 1]
```diff
@@ -61,6 +61,17 @@ else
 fi
 
 mke2fs -Fjqt $OSP_FILESYSTEM $filesystem_dev
+
+# in some cases, mkfs will cause the partition tables to remap, for example
+# /dev/mapper/foo-1 will turn into /dev/mapper/foo-part1. Rebuild the map
+# to ensure the filesystem can be detected onward:
+if [ ! -b "$filesystem_dev" ]; then
```
Comment (SuperQ) [inline create]: Minor bash style nit, it's better to use `[[ ]]` tests in bash.

**Review by SuperQ — APPROVED** [review]

File: create (hunk 1 above)
Comment (anarcat) [reply create]: elsewhere the code doesn't use `[[ ]]`, so it would look odd. besides: why?

Comment (SuperQ) [reply create]: There are some surprising gotchas with single brackets.

https://www.delftstack.com/howto/linux/difference-between-single-and-double-square-brackets-in-bash/

**Comment by anarcat** [comment]
On 2022-05-11 13:55:45, Ben Kochie wrote:
> There are some surprising gotchas with single brackets.
>
> https://www.delftstack.com/howto/linux/difference-between-single-and-double-square-brackets-in-bash/

Can you outline a specific problem you can see in that page? i don't see
much apart from && and || not being supported in plain "test" which is
fine by me (you just use the normal shell operators instead).

in general, i think we should strive to avoid bashism, and, in this
specific case, we would add one where there wasn't any before. all other
uses of `test` in that file use single brackets, why focus on this one?

**Comment by SuperQ** [comment]
I'm just saying, if it's a new bit of code, following bash best practices is a good idea. I also think it's fine as-is.

**Comment by saschalucas** [comment]
Thanks to everyone stressing the point, that there are issues regarding partition mapping. However the reason for seems not clear:

@anarcat states in #13 that he is using `blockdev` disk template in combination with iSCSI. It is common to use iSCSI with multipath. From `multipath.conf(5)` `skip_kpartx` defaults to `no`. So there's a chance that two different `kpartx` fight against each other?

Does @SuperQ also use something around possible multipath, or is it just `plain`/`drbd` etc. disk template?

**Comment by jcharaoui** [comment]
> @anarcat states in #13 that he is using `blockdev` disk template in combination with iSCSI. It is common to use iSCSI with multipath. From `multipath.conf(5)` `skip_kpartx` defaults to `no`. So there's a chance that two different `kpartx` fight against each other?

It's possible that it could work with `skip_kpartx=no`, I haven't tested this. However this is somewhat besides the point, because some of our machines *need* those partitions to be exposed by multipath, otherwise our setup will break and these machines will fail to find their `blockdev` on the next reboot of the nodes.

Creating a new instance should work regardless of `skip_kpartx`, and it can be done rather easily.

File: create [hunk 2]
```diff
@@ -61,6 +61,17 @@ else
 fi
 
 mke2fs -Fjqt $OSP_FILESYSTEM $filesystem_dev
+
```
Comment (anarcat) [inline create]: testing showed that this won't work because it's *after* the `mk2fs` call. it should at least be done *before*.

**Review by anarcat — CHANGES_REQUESTED** [review]

**Comment by anarcat** [comment]
> Thanks to everyone stressing the point, that there are issues regarding partition mapping. However the reason for seems not clear:
> 
> @anarcat states in #13 that he is using `blockdev` disk template in combination with iSCSI. It is common to use iSCSI with multipath. From `multipath.conf(5)` `skip_kpartx` defaults to `no`. So there's a chance that two different `kpartx` fight against each other?
> 
> Does @SuperQ also use something around possible multipath, or is it just `plain`/`drbd` etc. disk template?

i think you might have a very good point here. in fact, testing our procedures with this patch showed that it didn't fundamentally solve our problem. furthermore, looking again at #13, i see that I created multiple LUNs per VM, which is not something we do anymore. our new procedure now creates a single LUN, partitions it correctly, then passes those devices *individually* to the VM, with separate `adopt` blocks. 

It's not ideal, but it works, and I suspect it would work even without this patch. (but that remains to be tested.)

**Comment by saschalucas** [comment]
Thanks at @jcharaoui and @anarcat. I see now your point, that you somehow rely on multipath and partitions on this LUNs. Is this a nested stack of partition inside an other partition, or just two different mappings of the same partition (one side from multipath, the other from instance-debootstrap)? If the later is the case, I could image to align instance-debootstraps partition prefix with multipath, like in commit 90388ca (ATM overwritten by fore push), if that would solve your problem?

**Comment by jcharaoui** [comment]
@saschalucas They are different mappings of the same partitions, yes. The commit that I pushed and removed from this PR, the one that adjusts the partition delimiter, is what we ended up using to successfully. I verified that it also works when create DRBD instances as well. If that looks like a proper fix to you, you may close this PR and I'll submit a new one with that fix.

Also FTR I did try to adjust `partition_delimiter` in multipath.conf but did not find that it improved or even changed anything.

**Comment by saschalucas** [comment]
> If that looks like a proper fix to you, you may close this PR and I'll submit a new one with that fix.

Yes, I'm fine with changing the partition prefix in `instance-debootstrap` to be aligned with multipath.

I'm not sure about the right etiquette, but instead of closing this PR and creating a new one, we can change the title of this and force-push a different solution (namely commit 90388ca)? As PR creator @jcharaoui should be able to change title/close etc. If not I'll help out. Please decide for yourself which way you want to go.

**Comment by jcharaoui** [comment]
Alright @saschalucas, I've changed this PR according to what we discussed.

**Review by saschalucas — APPROVED** [review]
Thanks @jcharaoui. LGTM.

**Comment by saschalucas** [comment]
Just for reference: This aligns with https://github.com/opensvc/multipath-tools/blob/686797b7d0da4fc24fdfcb0fc6364df78a0b232d/kpartx/kpartx.rules#L42

**Pull request merged**